### PR TITLE
Properly compare unequal function pointers

### LINKED ISSUE #577: Better error messages when comparing fn ptrs
The following code

```rust
fn f() -> i32 {
    42
}

fn return_fn_ptr() -> fn() -> i32 {
    f
}

fn main() {
    assert!(return_fn_ptr() != f);
}
```

results in the following error:

```
error[E0080]: constant evaluation error: tried to dereference a function pointer
  --> fnptr.rs:10:13
   |
10 |     assert!(return_fn_ptr() == f);
   |             ^^^^^^^^^^^^^^^^^^^^ tried to dereference a function pointer
   |
   = note: inside call to `main` at /home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:74:34
   = note: inside call to closure at /home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:62:53
   = note: inside call to closure at /home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panicking.rs:307:40
   = note: inside call to `std::panicking::try::do_call::<[closure@DefId(1/1:1898 ~ std[8e5f]::rt[0]::lang_start_internal[0]::{{closure}}[0]) 0:&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe], i32>` at /home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panicking.rs:303:5
   = note: inside call to `std::panicking::try::<i32, [closure@DefId(1/1:1898 ~ std[8e5f]::rt[0]::lang_start_internal[0]::{{closure}}[0]) 0:&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe]>` at /home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panic.rs:398:9
   = note: inside call to `std::panic::catch_unwind::<[closure@DefId(1/1:1898 ~ std[8e5f]::rt[0]::lang_start_internal[0]::{{closure}}[0]) 0:&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe], i32>` at /home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:62:25
   = note: inside call to `std::rt::lang_start_internal` at /home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:74:5
   = note: inside call to `std::rt::lang_start::<()>`
```

This error message is factually incorrect, the function pointer does not get dereferenced.  We should fix that, either by making the test case pass or by at least giving a more accurate error message.

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #577

### CODE REVIEW

File: tests/run-pass/fn_ptr_comparison.rs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+fn f() -> i32 {
+    42
+}
+
+fn return_fn_ptr() -> fn() -> i32 {
+    f
+}
+
+fn main() {
+    assert!(return_fn_ptr() != f);
```
Comment (RalfJung) [inline tests/run-pass/fn_ptr_comparison.rs]: Can you add this to the existing fn ptr test instead?

File: tests/compile-fail/fn_ptr_comparison.rs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+fn f() -> i32 {
+    42
+}
+
+fn return_fn_ptr() -> fn() -> i32 {
+    f
+}
+
+fn main() {
+    assert!(return_fn_ptr() == f); //~ ERROR assertion failed
```
Comment (RalfJung) [inline tests/compile-fail/fn_ptr_comparison.rs]: What is the purpose of this test? We already test elsewhere that `assert!(false)` fails, don't we?

Comment (oli-obk) [reply tests/compile-fail/fn_ptr_comparison.rs]: I just wanted to make sure we're getting the right error. But I guess the run-pass test kinda guarantees that

File: src/operator.rs [hunk 3]
```diff
@@ -143,12 +143,24 @@ impl<'a, 'mir, 'tcx> EvalContextExt<'tcx> for super::MiriEvalContext<'a, 'mir, '
...
                     // Dead allocations in miri cannot overlap with live allocations, but
-                    // on read hardware this can easily happen. Thus for comparisons we require
+                    // on real hardware this can easily happen. Thus for comparisons we require
                     // both pointers to be live.
-                    self.memory().get(left.alloc_id)?.check_bounds_ptr(left)?;
-                    self.memory().get(right.alloc_id)?.check_bounds_ptr(right)?;
-                    // Two in-bounds pointers, we can compare across allocations
-                    left == right
+
+                    let check = |ptr: Pointer<Borrow>| match self.memory().get(ptr.alloc_id) {
```
Comment (RalfJung) [inline src/operator.rs]: This is not nice. :/

For once, it's not correct: A pointer into a function allocation is only in-bounds if its offset is 0.

But also, the previous behavior that `memory().check_bounds` would accept offset-0-size-0 pointers into function allocations was fully intended. These pointers *are* in-bounds. There is nothing wrong with a `&()` pointing to a function allocation. So I don't think fixing this here is how we want to proceed.

For example, `pointer_offset_inbounds` now also fails to compute a 0-offset on a function pointer. This affects every user of the old `check_bounds` method that was converted to `memory().get`.

Comment (oli-obk) [reply src/operator.rs]: I can reintroduce the `check_bounds` method on `Memory`

Comment (RalfJung) [reply src/operator.rs]: Yeah I think that would make most sense.

**Comment by RalfJung** [comment]
Superseded by https://github.com/solson/miri/pull/587

**Pull request closed without merging**